src/buildHTML.js

```diff
@@ -206,166 +206,13 @@ export const isLeftTight = function(node) {
-        const marginRight = (0.5 / metrics.ptPerEm) / multiplier + "em";
-
-        let supsub;
-        if (!group.value.sup) {
-            // Rule 18b
-            subShift = Math.max(
-                subShift, metrics.sub1,
-                subm.height - 0.8 * metrics.xHeight);
-
-            const vlistElem = [{type: "elem", elem: subm, marginRight}];
-            // Subscripts shouldn't be shifted by the base's italic correction.
-            // Account for that by shifting the subscript back the appropriate
-            // amount. Note we only do this when the base is a single symbol.
-            if (base instanceof domTree.symbolNode) {
-                vlistElem[0].marginLeft = -base.italic + "em";
-            }
-
-            supsub = buildCommon.makeVList({
-                positionType: "shift",
-                positionData: subShift,
-                children: vlistElem,
-            }, options);
-        } else if (!group.value.sub) {
-            // Rule 18c, d
-            supShift = Math.max(supShift, minSupShift,
-                supm.depth + 0.25 * metrics.xHeight);
-
-            supsub = buildCommon.makeVList({
-                positionType: "shift",
-                positionData: -supShift,
-                children: [{type: "elem", elem: supm, marginRight}],
-            }, options);
-        } else {
-            supShift = Math.max(
-                supShift, minSupShift, supm.depth + 0.25 * metrics.xHeight);
-            subShift = Math.max(subShift, metrics.sub2);
-
-            const ruleWidth = metrics.defaultRuleThickness;
-
-            // Rule 18e
-            const maxWidth = 4 * ruleWidth;
-            if ((supShift - supm.depth) - (subm.height - subShift) < maxWidth) {
-                subShift = maxWidth - (supShift - supm.depth) + subm.height;
-                const psi = 0.8 * metrics.xHeight - (supShift - supm.depth);
-                if (psi > 0) {
-                    supShift += psi;
-                    subShift -= psi;
-                }
-            }
-
-            const vlistElem = [
-                {type: "elem", elem: subm, shift: subShift, marginRight},
-                {type: "elem", elem: supm, shift: -supShift, marginRight},
-            ];
-            // See comment above about subscripts not being shifted.
-            if (base instanceof domTree.symbolNode) {
-                vlistElem[0].marginLeft = -base.italic + "em";
-            }
-
-            supsub = buildCommon.makeVList({
-                positionType: "individualShift",
-                children: vlistElem,
-            }, options);
-        }
-
-        // Wrap the supsub vlist in a span.msupsub to reset text-align.
-        const mclass = getTypeOfDomTree(base) || "mord";
-        return makeSpan([mclass],
-            [base, makeSpan(["msupsub"], [supsub])],
-            options);
-    },
-};
+/** This is a map of group types to the function used to handle that type. */
+export const groupTypes = {};
```

Comment:
Nice!
What do you think about moving them to `src/defineFunction.js` like `_functions`?

Reply:
Honestly, not sure at all. Currently, `groupTypes` are used by both the `build*.js` files and the `defineFunction` and `defineEnvironment`. I honestly can't tell the difference/advantage in moving one place or the other. The only minor consideration that comes to mind is that if we move both `groupTypes` into `defineFunction`, we're going to have to come up with two different names for them since they would be in the same file, whereas now they are accessed as `mml.groupTypes` and `html.groupTypes` where they are used, which is kinda nice.

They really should be imported nowhere else, though I see that `functions/raisebox` requires it -- I can clean that up by having it require `size`'s `buildBuilder` directly. I'd do it in a separate PR, but it can also be in this one if you prefer.

Reply:
Ah, it's also used in `defineEnvironment`. I though they are only defined in `defineFunction`.

Reply:
@marcianx Too late but what do you think about renaming them to `groupBuilders`?

Reply:
You know what, I'm definitely empathizing with this now (see the `raisebox` PR I'll send soon). All the `functions/*` generally import *all* the exports within `buildHTML` and `buildMathML`, but they really should not be accessing `groupTypes` at all. So you're right. I'll move this into `src/defineFunction` and not import them anywhere else.